syntax = "proto3";

package bittorrent;

// =============================================================
// SERVICIO 1: El Tracker (El "Directorio")
// =============================================================
service TrackerService {
    // Se usa al inicio y periódicamente (heartbeat)
    rpc RegistrarNodo (InfoNodo) returns (AckTracker);
    rpc BuscarArchivo (BusquedaArchivo) returns (ListaPeers);
    
    // (Opcional) Para ver la lista global en la consola del Tracker (punto 2 del PDF)
    rpc ObtenerEstadoSwarm (Vacio) returns (EstadoSwarm);
}

// =============================================================
// SERVICIO 2: El Nodo P2P (Transferencia de archivos)
// =============================================================
service P2PService {
    rpc SolicitarChunk (PeticionChunk) returns (DataChunk);
    
    // NUEVO: Preguntar metadatos del archivo, para que sepa cuánto pesan los archivos
    rpc ObtenerInfoArchivo (BusquedaArchivo) returns (InfoArchivo);
}

// =============================================================
// MENSAJES (Las estructuras de datos)
// =============================================================

message Vacio {}

message InfoNodo {
    string ip_puerto = 1;       // Ejemplo: "192.168.1.50:50051"
    repeated string archivos = 2; // Lista de archivos completos que tiene: ["video.mp4", "t.txt"]

}

message AckTracker {
    bool exito = 1;
    string mensaje = 2;
}

message BusquedaArchivo {
    string nombre_archivo = 1;
}

message ListaPeers {
    // Devuelve una lista de direcciones IP:Puerto que tienen el archivo
    repeated string peers_con_archivo = 1; 
}

message PeticionChunk {
    string nombre_archivo = 1;
    int32 indice_chunk = 2;   // ¿Qué numero de pedazo quieres? (0, 1, 2...)
}

message DataChunk {
    bool encontrado = 1;
    bytes datos = 2;          
    int32 indice_chunk = 3;   
}

message EstadoSwarm {
    repeated InfoNodo nodos = 1;
}

message InfoArchivo {
    string nombre = 1;
    int64 tamano_bytes = 2;   // Usamos int64 para archivos grandes
    int32 total_chunks = 3;   // Cuántos pedazos son en total
    bool existe = 4;
}
